codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3UpdateEndpoint.java

```diff
@@ -66,17 +81,45 @@ public void writeAdditionalFiles(
             GoDelegator goDelegator
     ) {
         ServiceShape service = settings.getService(model);
-        if (!isS3Service(model, service)) {
-            return;
+
+        // if service is s3control
+        if (isS3ControlService(model, service)){
+            goDelegator.useShapeWriter(service, this::writeS3ControlMiddlewareHelper);
         }
 
-        goDelegator.useShapeWriter(service, this::writeMiddlewareHelper);
+        // check if service is s3
+        if (isS3Service(model, service)) {
+            goDelegator.useShapeWriter(service, this::writeS3MiddlewareHelper);
+
+            goDelegator.useShapeWriter(service, writer -> {
+                writeInputGetter(writer, model, symbolProvider, service);
+            });
+
+            goDelegator.useShapeWriter(service, writer -> {
+                writeAccelerateValidator(writer, model, symbolProvider, service);
+            });
+        }
+    }
 
-        goDelegator.useShapeWriter(service, writer -> {
-            writeInputGetter(writer, model, symbolProvider, service);
+    private void writeAccelerateValidator(GoWriter writer, Model model, SymbolProvider symbolProvider, ServiceShape service) {
+        writer.writeDocs("supportAccelerate returns a boolean indicating if the operation associated with the provided input "
+                + "supports S3 Transfer Acceleration");
+        writer.openBlock("func $L(input interface{}) bool {", "}", SUPPORT_ACCELERATE, () -> {
+            writer.openBlock("switch input.(type) {" , "}", () -> {
+                for (ShapeId operationId : service.getAllOperations()) {
+                    // check if operation does not support s3 accelerate
+                    if (NOT_SUPPORT_ACCELERATE.contains(operationId.getName())) {
+                        OperationShape operation = model.expectShape(operationId, OperationShape.class);
+                        StructureShape input = model.expectShape(operation.getInput().get(), StructureShape.class);
+                        writer.write("case $P: return false", symbolProvider.toSymbol(input));
+                    }
```

Comment:
AT some point we should consider if its better to have this large switch statements, or individual functions per operation?  The benefit of the individual function is that there is no switch statements. I doubt large switch statements are cheap.

Reply:
#847 will take care of this